istioctl/pkg/auth/upgrader.go

```diff
@@ -140,9 +176,18 @@ func getAuthorizationPolicies(policies []model.Config) (*model.AuthorizationPoli
 
 // convert is the main function that converts RBAC v1alphal1 to v1beta1 for local policy files
 func (ug *Upgrader) convert(authzPolicies *model.AuthorizationPolicies) error {
+	// Convert ClusterRbacConfig to AuthorizationPolicy
+	err := ug.convertClusterRbacConfig(authzPolicies)
+	if err != nil {
+		// Users might not have access to the cluster-wide RBAC config, so instead of returning an error,
+		// output a warning instead.
+		log.Warnf("failed to convert ClusterRbacConfig: %s", err)
+	}
 	namespaces := authzPolicies.ListNamespacesOfToV1alpha1Policies()
 	if len(namespaces) == 0 {
-		return fmt.Errorf("no namespace found from the provided authorization policy")
+		// Similarly, a user might want to convert ClusterRbacConfig only.
+		log.Warn("no namespace found for ServiceRole and ServiceRoleBinding")
+		return nil
```

Comment:
If this is printed to stdout won't it conflict with the successfully converted policy? Can you double check and ensure this goes to stderr?

Reply:
These logs will always get printed first before the converted policy gets printed out, so I think it's fine?

Reply:
IIUC, the expected usage for this command would be something like:

```bash
istioctl x auth upgrade old-policy.yaml > upgraded-policy.yaml
kubectl apply -f upgraded-policy.yaml
```
If so, the output of `upgrade` must be a valid YAML document. Printing the warning at all to stdout will intermix the log message with the YAML content. You'll need to either log to stderr or prefix warnings with `#` so that they appear as comments in the yaml output.

Reply:
That's a valid concern, but I think (1): the users should inspect the output of the upgraded policy before applying it, (2): more importantly, if there is some error (or warning such as ClusterRbacConfig doesn't exist - this is not necessarily an error, maybe this is what the users want), the kubectl apply will be rejected at Galley admission webhook. This can further prevent errors if the users blindly apply upgraded policy without inspecting it first. I think each approach has its pros and cons, but at least the current approach is safer. Wdyt? Another point I mentioned above is that these logs (two at most) will be printed out first, so if the policy looks right to the users, they can simply remove them at the beginning of their files.

Reply:
Is there any reason we can't print to stderr? The user would see the error message even if they were piping output to a file.

Reply:
Sorry, that comment above was for the adding `#` approach. As for the stderr, is there a benefit in this case? I'd imagine a user would run istioctl cmd ..., so stuff from stderr and stdout will be shown in stdout?

If there is a strong reason we should do stderr in this case, is there a strong preference how we should do it? e.g. `l := log.New(os.Stderr, "", 0)`, `fmt.Fprintf`, etc.

Reply:
stderr and stdout are both printed to the terminal but they are separate streams. `istioctl x auth upgrade missing-namespace.yaml > upgraded.yaml` redirects stdout to the file but stderr will still print to the terminal.

The warning is a user facing message and not really an internal log. I would use cmd.OutOrStderr() to create a stderr writer and pass that to the upgrader.

Reply:
That makes sense. I pushed a commit to reflect the change. PTAL.

Reply:
If you must log, we no longer get complaints about log messages in kubeinject.go: https://github.com/istio/istio/blob/master/istioctl/cmd/kubeinject.go#L312
which causes the logs to write to stderr by default.  It is an easy fix and the user can override it.

Reply:
That looks great. Wdyt @ayj?

Reply:
I'm leaning toward Ed's suggestion. It also reduces the burden of changing the existing method of how we're testing `auth x` commands.

Reply:
That's fine. It is a little weird to see time/level prefixed messages in the default CLI output, but we seem to be doing that already in several other places it seems.

Reply:
@esnible The kube-inject command didn't seem to work (setting default log to stderr) when I tried it. If I add a log, it still gets directed to stdout, instead of stderr. I had to manually set `--log_target stderr` to make it work.

Reply:
False alarm, looks like it's not broken.